Added Banned Identifiers

DanielEasson opened the pull request:
This pull request addresses the recent CI failure by banning some identifiers from being used in the generation process. This is based on a very rare case where the generated identifiers share a name with an export from the runtime and compiletime libraries. If this happens, an Ambiguous Occurrence error will be given, failing the test. Once again my apologies for the CI failure, these types of bugs recently appearing never occurred when I ran the tests locally, but I am glad they have been identified early.

DanielEasson committed:
Add test suite for BNFC-meta

ulysses4ever committed:
Merge branch 'master' into master

DanielEasson committed:
Delete BNFC-meta/test/Grammars/FailureGrammars.hs

Extra file for testing that is no longer needed.

DanielEasson committed:
Fixed Numerous Level Three Bugs

DanielEasson committed:
Fixed Numerous Level Three Bugs

DanielEasson committed:
Removed Unnecessary Prints

DanielEasson committed:
Specify version bounds following PVP for dependencies

DanielEasson committed:
Update BNFC-meta.cabal

Fixed boundaries

DanielEasson committed:
Update Instances.hs

Added markers to prevent rare ambiguity case level three grammar generation.

DanielEasson committed:
Merge branch 'ulysses4ever:master' into master

DanielEasson committed:
Update Instances.hs

Resized level two LBNF test as there was a very rare chance, the generated parsetree would be too large.

DanielEasson committed:
Merge branch 'ulysses4ever:master' into master

DanielEasson committed:
Update Instances.hs

Added banned Identifiers for the incredibly rare chance that the generated data structure shares names with the exports from compiletime and runtime.

ulysses4ever reviewed (approved):
Looks good, thanks!

For future PRs it'd easier to handle them if you created new branch from the latest matter every time. In that case I'd avoid squashing your old commits and editing commit messages every time. But that's not big deal of course.

ulysses4ever commented:
The solution with banned lists is fragile. A better approach would be to generate a random identifier and add a constant prefix that would guarantee to disambiguate,e.g. ` ++ "bnfcMetaTest"`. Do you think you could implement that?

DanielEasson commented:
Yes, that's a much better approach.

DanielEasson committed:
Update Instances.hs

Added the fixed prefix BnfcMetaTest, instead of checking for banned identifiers
